File: Data Structures & Algorithms/max-water-container/submission-1.cpp
class Solution {
public:
    int maxArea(vector<int>& height) {
        int l = 0;
        int r = height.size() - 1;

        int area = 0;

        while (l < r) {
            area = max((r - l) * min(height[l], height[r]), area);

            if (height[l] < height[r]) {
                l++;
            } else {
                r--;
            }
        }

        return area;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-1.py
class Solution:
    def trap(self, height: List[int]) -> int:
        
        l, r = 0, len(height) - 1

        mh = min(height[l], height[r])

        w = 0

        while l < r:
            mh = max(mh, min(height[l], height[r]))

            if height[l] < height[r]:
                w += max(0, mh - height[l])
                l += 1
            else:
                w += max(0, mh - height[r])
                r -= 1

        return w